# Week 1 Checkpoint Assignment: Data Preprocessing with pandas

**Name:** _Joe Doan______________________  
**Student ID:** _16389776________________  
**Date:** ____15 Sep_____________________

**Points:** 60  
**Suggested time:** 60–75 minutes

This is an individual end-of-session exercise. It assesses your ability to inspect, filter, clean, and summarize tabular data with pandas.

## Submission requirements

Submit both of the following:

1. Your completed notebook, renamed **Lastname_Firstname_Week1_Checkpoint.ipynb**.
2. The cleaned CSV created in Task 7, named **hepatitis_cleaned.csv**.

Before submitting:

- run the notebook from top to bottom;
- resolve all error messages;
- keep your outputs visible;
- answer every written-response prompt in a Markdown cell;
- do not modify the original dataset file.

You may consult pandas documentation and course notes. Your submitted code and explanations must be your own.

## Dataset and allowed tools

Use the hepatitis dataset supplied with the lesson. The notebook loads the local copy first and uses the public source only as a fallback.

You may use standard pandas and NumPy methods covered in class, including:

- <code>head</code>, <code>shape</code>, <code>info</code>, and <code>describe</code>;
- column selection and <code>loc</code>;
- Boolean filtering;
- <code>select_dtypes</code> and <code>value_counts</code>;
- <code>isna</code>, <code>dropna</code>, and <code>fillna</code>;
- <code>duplicated</code>, <code>drop_duplicates</code>, and <code>reset_index</code>;
- <code>groupby</code>, <code>sort_values</code>, and <code>to_csv</code>.

## Setup

Run this cell without changing it.

In [37]:
from pathlib import Path
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", 25)

local_path = Path("data/hepatitis.csv")
data_url = "https://raw.githubusercontent.com/datasets/hepatitis/master/data/hepatitis.csv"
source = local_path if local_path.exists() else data_url

df_raw = pd.read_csv(source)
print(f"Dataset loaded from: {source}")

Dataset loaded from: https://raw.githubusercontent.com/datasets/hepatitis/master/data/hepatitis.csv


## Task 1 — Confirm the dataset (4 points)

1. Display the first **seven** rows.
2. Store the dataset dimensions in a variable named <code>dataset_shape</code>.
3. Print a sentence reporting the number of rows and columns.

Your code must use the DataFrame rather than typing the dimensions manually.

In [38]:
# Task 1 solution
df_raw.head(7)

,age,sex,steroid,antivirals,fatigue,malaise,anorexia,liver_big,liver_firm,spleen_palpable,spiders,ascites,varices,bilirubin,alk_phosphate,sgot,albumin,protime,histology,class
0,30,male,False,False,False,False,False,False,False,False,False,False,False,1.0,85.0,18.0,4.0,NaN,False,live
1,50,female,False,False,True,False,False,False,False,False,False,False,False,0.9,135.0,42.0,3.5,NaN,False,live
2,78,female,True,False,True,False,False,True,False,False,False,False,False,0.7,96.0,32.0,4.0,NaN,False,live
3,31,female,NaN,True,False,False,False,True,False,False,False,False,False,0.7,46.0,52.0,4.0,80.0,False,live
4,34,female,True,False,False,False,False,True,False,False,False,False,False,1.0,NaN,200.0,4.0,NaN,False,live
5,34,female,True,False,False,False,False,True,False,False,False,False,False,0.9,95.0,28.0,4.0,75.0,False,live
6,51,female,False,False,True,False,True,True,False,True,True,False,False,NaN,NaN,NaN,NaN,NaN,False,die


In [39]:
dataset_shape = df_raw.shape

print(f'Data shape: {dataset_shape}')

Data shape: (155, 20)


## Task 2 — Inspect structure and quality (8 points)

1. Display the DataFrame's concise structural summary.
2. Store the data type of every column in <code>column_types</code>.
3. Create <code>numeric_summary</code>, containing descriptive statistics for numeric columns.
4. In the Markdown cell below, identify:
   - one numeric variable;
   - one categorical variable;
   - one column that contains missing values.

Include evidence from your output.

In [40]:
#1 Display the DataFrame's concise structural summary.
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              155 non-null    int64  
 1   sex              155 non-null    object 
 2   steroid          154 non-null    object 
 3   antivirals       155 non-null    bool   
 4   fatigue          154 non-null    object 
 5   malaise          154 non-null    object 
 6   anorexia         154 non-null    object 
 7   liver_big        145 non-null    object 
 8   liver_firm       144 non-null    object 
 9   spleen_palpable  150 non-null    object 
 10  spiders          150 non-null    object 
 11  ascites          150 non-null    object 
 12  varices          150 non-null    object 
 13  bilirubin        149 non-null    float64
 14  alk_phosphate    126 non-null    float64
 15  sgot             151 non-null    float64
 16  albumin          139 non-null    float64
 17  protime         

In [41]:
#2 Store the data type of every column
column_types = df_raw.dtypes
column_types

,0
age,int64
sex,object
steroid,object
antivirals,bool
fatigue,object
malaise,object
anorexia,object
liver_big,object
liver_firm,object
spleen_palpable,object


In [42]:
#3 Create numeric_summary, containing descriptive statistics for numeric columns.
numeric_columns = ['age', 'bilirubin', 'alk_phosphate', 'sgot', 'albumin', 'protime']
numeric_summary = df_raw[numeric_columns].describe()
numeric_summary

,age,bilirubin,alk_phosphate,sgot,albumin,protime
count,155.000000,149.000000,126.000000,151.00000,139.000000,88.000000
mean,41.200000,1.427517,105.325397,85.89404,3.817266,61.852273
std,12.565878,1.212149,51.508109,89.65089,0.651523,22.875244
min,7.000000,0.300000,26.000000,14.00000,2.100000,0.000000
25%,32.000000,0.700000,74.250000,31.50000,3.400000,46.000000
50%,39.000000,1.000000,85.000000,58.00000,4.000000,61.000000
75%,50.000000,1.500000,132.250000,100.50000,4.200000,76.250000
max,78.000000,8.000000,295.000000,648.00000,6.400000,100.000000


* numberic variable: ['age', 'bilirubin', 'alk_phosphate', 'sgot', 'albumin', 'protime']

* categorical variable: ['antivirals', 'histology']

* one column that contains missing values:

In [43]:
df_raw[df_raw['liver_firm'].isna()].head(1)

,age,sex,steroid,antivirals,fatigue,malaise,anorexia,liver_big,liver_firm,spleen_palpable,spiders,ascites,varices,bilirubin,alk_phosphate,sgot,albumin,protime,histology,class
31,62,female,False,False,True,True,False,NaN,NaN,False,False,False,False,1.0,NaN,60.0,NaN,NaN,False,die


### Task 2 written response

1. One number variable: age, bilirubin, alk_phosphate,..
* Evidence: In df_raw.info() table, they're
stated as int64 and float64 dtypes.
---
2. One Categorical Variable: sex, steroid, antivirals,...
* Evidence: In df_raw.info() table, they're stated as boolean or object dtypes, which is represents textual or categorical categories in pandas
---
3. One Columnn with missing values: protime,alk_phosphate,...
* Evidence: df_raw.info() reports a total of 155 indices, but these columns store less than 155 non-null counts.



## Task 3 — Select and filter data (10 points)

### Part A

Create <code>selected_columns</code> containing only these columns, in this order:

- age
- sex
- albumin
- bilirubin
- class

Display its first five rows.

### Part B

Create <code>filtered_patients</code> containing patients who:

- are younger than 30; **and**
- have an albumin value greater than or equal to 4.

Keep only <code>age</code>, <code>albumin</code>, and <code>class</code>. Sort the result by age from youngest to oldest.

In [44]:
# Task 3 solution
selected_columns = df_raw[['age', 'sex', 'albumin', 'bilirubin', 'class']]
selected_columns.head(5)

,age,sex,albumin,bilirubin,class
0,30,male,4.0,1.0,live
1,50,female,3.5,0.9,live
2,78,female,4.0,0.7,live
3,31,female,4.0,0.7,live
4,34,female,4.0,1.0,live


In [45]:
filtered_patients = df_raw[(df_raw['age'] < 30) & (df_raw['albumin'] >= 4)]
filtered_patients = filtered_patients[['age', 'albumin', 'class']]
filtered_patients.sort_values('age')

,age,albumin,class
110,7,4.2,live
148,20,4.0,live
20,22,4.2,live
41,23,4.6,live
80,23,4.8,live
36,23,4.1,live
70,24,4.3,live
116,24,4.1,live
135,25,4.5,live
105,25,6.4,live


In [46]:
filtered_patients.shape

(17, 3)

### Task 3 written response

How many rows meet both filtering conditions? Explain how you obtained the answer.

There are 17 tows meet both filtering conditions. I used shape function to obtain the answer.


## Task 4 — Explore categorical variables (8 points)

1. Create a list named <code>categorical_columns</code> using pandas data-type selection. Do not type the column names manually.
2. Create <code>steroid_counts</code> showing every value in the steroid column, including missing values.
3. Display the number of unique non-missing values in every categorical column.
4. Briefly explain why including missing values in a frequency table can be useful.

In [47]:
# Task 4 solution
categorical_columns = df_raw.select_dtypes(include=['object', 'bool']).columns.tolist()
categorical_columns

['sex',
 'steroid',
 'antivirals',
 'fatigue',
 'malaise',
 'anorexia',
 'liver_big',
 'liver_firm',
 'spleen_palpable',
 'spiders',
 'ascites',
 'varices',
 'histology',
 'class']

In [48]:
steroid_counts = df_raw['steroid'].value_counts(dropna=False)
steroid_counts

,count
steroid,
True,78
False,76
NaN,1


### Task 4 written response

We add missing values to the frequency table because it immediately shows the data quality of that category and ensures the total sum of frequencies matches the entire dataset's size.


## Task 5 — Analyze missing values (12 points)

Create a DataFrame named <code>missing_summary</code> with:

- one row per original column;
- a column named <code>missing_count</code>;
- a column named <code>missing_percent</code>, rounded to one decimal place;
- rows sorted from highest to lowest missing percentage.

Then:

1. display only columns with at least one missing value;
2. store the name of the column with the most missing values in <code>highest_missing_column</code>;
3. calculate how many rows would remain after dropping every row containing at least one missing value, and store the result in <code>complete_row_count</code>.

In [49]:
# Task 5 solution
missing_summary = df_raw.isna().sum().to_frame(name='missing_count')
missing_summary['missing_percent'] = (missing_summary['missing_count'] / len(df_raw) * 100).round(1)

missing_summary = missing_summary.sort_values('missing_percent', ascending=False)
missing_summary

,missing_count,missing_percent
protime,67,43.2
alk_phosphate,29,18.7
albumin,16,10.3
liver_firm,11,7.1
liver_big,10,6.5
bilirubin,6,3.9
spleen_palpable,5,3.2
spiders,5,3.2
varices,5,3.2
ascites,5,3.2


In [50]:
highest_missing_column = missing_summary.index[0]
missing_summary[missing_summary['missing_count'] != 0]

,missing_count,missing_percent
protime,67,43.2
alk_phosphate,29,18.7
albumin,16,10.3
liver_firm,11,7.1
liver_big,10,6.5
bilirubin,6,3.9
spleen_palpable,5,3.2
spiders,5,3.2
varices,5,3.2
ascites,5,3.2


In [51]:
highest_missing_column

'protime'

In [52]:
complete_row_count = df_raw.dropna().shape[0]
complete_row_count

80

### Task 5 written response

Would dropping every incomplete row be a reasonable default decision for this dataset? Support your answer with the number and percentage of rows that would be lost.

Dropping every missing row will execute a big accuracy issue in analysis results, as we can observe there’s a column that has 43.2% of incomplete rows, so if we discard all of it, we will lose nearly half of our dataset volume. We can address it with other methods instead of dropping too much data.

## Task 6 — Build a cleaned copy (12 points)

Create <code>df_clean</code> as a copy of <code>df_raw</code>. Do not change <code>df_raw</code>.

Apply all of these rules to <code>df_clean</code>:

1. Fill missing numeric values with each column's median.
2. Fill missing non-numeric values with each column's mode.
3. Remove duplicate rows.
4. Reset the index after duplicate removal.

Finally, display:

- total missing values before and after cleaning;
- duplicate row count after cleaning;
- the shape of the cleaned DataFrame.

In [53]:
# Task 6 solution
df_clean = df_raw.copy()
numeric_fill_columns = df_clean.select_dtypes(include=['float64', 'int64']).columns
non_numeric_fill_columns = df_clean.select_dtypes(include=['object', 'bool']).columns
df_clean[numeric_fill_columns] = df_clean[numeric_fill_columns].fillna(df_clean[numeric_fill_columns].median())
df_clean[non_numeric_fill_columns] = df_clean[non_numeric_fill_columns].fillna(df_clean[non_numeric_fill_columns].mode().iloc[0])


/tmp/ipykernel_809/641277264.py:6: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_clean[non_numeric_fill_columns] = df_clean[non_numeric_fill_columns].fillna(df_clean[non_numeric_fill_columns].mode().iloc[0])


In [54]:
df_clean = df_clean.drop_duplicates().reset_index(drop=True)

print(f'Total missing values before cleaning: {df_raw.isna().sum().sum()}')
print(f'Total missing values after cleaning: {df_clean.isna().sum().sum()}')
print(f'Duplicate row count after cleaning: {df_raw.duplicated().sum()}')
print(f'Shape of the cleaned DataFrame: {df_clean.shape}')

Total missing values before cleaning: 167
Total missing values after cleaning: 0
Duplicate row count after cleaning: 0
Shape of the cleaned DataFrame: (155, 20)


### Task 6 written response

Explain one advantage and one limitation of median/mode imputation.


* Median: Less sensitve to outliers, but it can distort the natural variance and distribution of continuos numerical data.
* Mode: Easy to apply to categorical data, but it can introduce bias by over-representing the most common category, especially if the dataset is small or unbalanced.

## Task 7 — Validate, summarize, and export (6 points)

1. Use assertions to confirm that:
   - <code>df_clean</code> has no missing values;
   - <code>df_clean</code> has no duplicate rows;
   - every age is between 0 and 120 inclusive.
2. Create <code>class_summary</code> containing the count and percentage of each value in the class column. Round percentages to one decimal place.
3. Export <code>df_clean</code> to **hepatitis_cleaned.csv** without saving the index.
4. In the Markdown cell below, explain why retaining <code>df_raw</code> is important.

In [55]:
# Task 7 solution
assert df_clean.isna().sum().sum() == 0
assert df_clean.duplicated().sum() == 0
assert df_clean['age'].between(0, 120).all()

In [56]:
class_summary = df_clean['class'].value_counts().to_frame(name='class_count')
class_summary['class_percentage'] = (class_summary['class_count'] / sum(class_summary['class_count']) * 100).round(1)
class_summary

,class_count,class_percentage
class,,
live,123,79.4
die,32,20.6


In [57]:
df_clean.to_csv('hepatitis_cleaned.csv', index=False)

### Task 7 written response

Keeping raw data is crucial because of its reproducibility, which allows us to audit the original data and try different cleaning strategies in the future.

## Submission self-check

Run this cell after completing all tasks. It checks required variable names and submission files. Passing this check does not guarantee full credit; code quality and written explanations are also graded.

In [58]:
required_variables = [
    "dataset_shape",
    "column_types",
    "numeric_summary",
    "selected_columns",
    "filtered_patients",
    "categorical_columns",
    "steroid_counts",
    "missing_summary",
    "highest_missing_column",
    "complete_row_count",
    "df_clean",
    "class_summary",
]

missing_variables = [
    name for name in required_variables if name not in globals()
]

if missing_variables:
    print("Still missing required variables:")
    for name in missing_variables:
        print(" -", name)
else:
    print("All required variables exist.")

output_file = Path("hepatitis_cleaned.csv")
print("Cleaned CSV exists:", output_file.exists())

if "df_raw" in globals():
    print("Raw dataset still contains missing values:", df_raw.isna().sum().sum() > 0)

if "df_clean" in globals():
    print("Clean dataset has zero missing values:", df_clean.isna().sum().sum() == 0)
    print("Clean dataset has zero duplicates:", not df_clean.duplicated().any())

All required variables exist.
Cleaned CSV exists: True
Raw dataset still contains missing values: True
Clean dataset has zero missing values: True
Clean dataset has zero duplicates: True


## Optional challenge — Reusable function (no extra credit required)

Write a function named <code>summarize_missing(data)</code> that accepts any DataFrame and returns a missing-value summary with counts and percentages, sorted from most to least missing.

Test the function using <code>df_raw</code>.

In [59]:
# Optional challenge
def summarize_missing(data):
  missing_summary = data.isna().sum().to_frame(name='missing_count')
  missing_summary['missing_percent'] = (missing_summary['missing_count'] / len(data) * 100).round(1)
  missing_summary = missing_summary.sort_values('missing_percent', ascending=False)
  return missing_summary

In [60]:
summarize_missing(df_raw)

,missing_count,missing_percent
protime,67,43.2
alk_phosphate,29,18.7
albumin,16,10.3
liver_firm,11,7.1
liver_big,10,6.5
bilirubin,6,3.9
spleen_palpable,5,3.2
spiders,5,3.2
varices,5,3.2
ascites,5,3.2


## Grading rubric

| Area | Points | Full-credit evidence |
|---|---:|---|
| Task 1: Load and confirm | 4 | Correct preview, programmatic shape, and clear printed result |
| Task 2: Inspect | 8 | Appropriate inspection methods, stored results, evidence-based explanation |
| Task 3: Select and filter | 10 | Correct columns, Boolean conditions, ordering, and row count |
| Task 4: Categories | 8 | Programmatic type selection, counts including missing values, explanation |
| Task 5: Missingness | 12 | Accurate count/percentage table, sorting, complete-row calculation, justified decision |
| Task 6: Cleaning | 12 | Copy preserved, correct imputation, duplicate handling, before/after evidence |
| Task 7: Validate and export | 6 | Assertions pass, class summary is correct, CSV exported, explanation completed |
| **Total** | **60** | |

Code should be readable, use meaningful variable names, and avoid unnecessary repetition.